api/lib/domain/services/recommendation-service.js

```diff
@@ -0,0 +1,67 @@
+const _ = require('lodash');
+
+const REACH_LEVEL_POINTS = 40;
+const DIFFICULTY_POINTS = 30;
+const PROGRESS_POINTS = 20;
+const FIRST_STEP_POINTS = 10;
+const DEFAULT_REACHED_LEVEL = 0;
+
+const FIRST_LEVEL = 1;
+
+function computeRecommendationScore(skills, maxSkillLevelInTargetProfile, knowledgeElements) {
+  const skill = _.maxBy(skills, 'difficulty');
+  const reachLevelInTube =  _getReachLevelInTube(knowledgeElements, skills);
+
+  const reachLevelScore = _computeReachedLevelScore(skill, reachLevelInTube);
+  const difficultyScore = _computeDifficultyScore(maxSkillLevelInTargetProfile, skill);
+  const progressScore = _computeProgressScore(skills, knowledgeElements);
+  const firstStepScore = _computeFirstStepScore(skills, knowledgeElements, reachLevelInTube);
+
+  return reachLevelScore + difficultyScore + progressScore + firstStepScore;
+}
+
+function _computeReachedLevelScore(skill, reachLevelInTube) {
+  return REACH_LEVEL_POINTS / skill.difficulty * reachLevelInTube;
+}
+
+function _computeProgressScore(skills, knowledgeElements) {
+  return PROGRESS_POINTS / skills.length * knowledgeElements.length;
+}
+
+function _computeFirstStepScore(skills, knowledgeElements, reachLevelInTube) {
+  const nextLevelToReach = _getNextLevelToReach(skills, knowledgeElements);
+
+  return FIRST_STEP_POINTS / nextLevelToReach * reachLevelInTube;
+}
+
+function _computeDifficultyScore(maxSkillLevelInTargetProfile, skill) {
+  return DIFFICULTY_POINTS / maxSkillLevelInTargetProfile * skill.difficulty;
+}
+
+function _getNextLevelToReach(skills, knowledgeElements) {
+  const nextSkillToAcquire = _(skills).filter((skill) => ! _.map(knowledgeElements, 'skillId').includes(skill.id))
+    .minBy('difficulty');
+
+  if (_tubeNotAssessed(knowledgeElements)) {
+    return FIRST_LEVEL;
+  } else if (!nextSkillToAcquire) {
+    return _.maxBy(skills, 'difficulty').difficulty;
+  }
+
+  return nextSkillToAcquire.difficulty;
+}
+
+function _tubeNotAssessed(knowledgeElements) {
```

Comment:
Un peu long, des idées ?

```suggestion
function _tubeWithoutValidatedKnowledgeElements(knowledgeElements) {
```

Reply:
On remplace par le _.isEmpty(validatedKnowledgeElements)